File: scripts/validate_json_files.js
#!/usr/bin/env node
// Walk the repository and ensure every JSON file parses correctly.

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT_DIR = path.resolve(__dirname, '..');
const IGNORE_DIRS = new Set(['.git', 'node_modules', '__pycache__']);

function findJsonFiles(dir) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const results = [];
  for (const entry of entries) {
    if (IGNORE_DIRS.has(entry.name)) {
      continue;
    }
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...findJsonFiles(fullPath));
    } else if (entry.isFile() && entry.name.endsWith('.json')) {
      results.push(fullPath);
    }
  }
  return results;
}

function validateFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  try {
    JSON.parse(content);
    return null;
  } catch (error) {
    return `${filePath}: ${error.message}`;
  }
}

function main() {
  const targets = findJsonFiles(ROOT_DIR);
  const errors = [];
  for (const file of targets) {
    const failure = validateFile(file);
    if (failure) {
      errors.push(failure);
    }
  }

  if (errors.length > 0) {
    console.error(`JSON validation failed for ${errors.length} file(s):`);
    for (const message of errors) {
      console.error(` - ${message}`);
    }
    process.exit(1);
  }

  console.log(`Validated ${targets.length} JSON file(s) successfully.`);
}

main();

File: scripts/validate_json_files.py
#!/usr/bin/env python3
"""Parse every JSON file in the repo to catch syntax errors early."""

import json
import os
import sys
from typing import List, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
IGNORE_DIRS = {".git", "node_modules", "__pycache__"}


def _find_json_files() -> List[str]:
    json_files: List[str] = []
    for dirpath, dirnames, filenames in os.walk(ROOT_DIR):
        dirnames[:] = [name for name in dirnames if name not in IGNORE_DIRS]
        for filename in filenames:
            if filename.endswith(".json"):
                json_files.append(os.path.join(dirpath, filename))
    return json_files


def _validate_files(paths: List[str]) -> List[Tuple[str, str]]:
    errors: List[Tuple[str, str]] = []
    for path in paths:
        try:
            with open(path, "r", encoding="utf-8") as handle:
                json.load(handle)
        except Exception as exc:  # pylint: disable=broad-except
            errors.append((path, str(exc)))
    return errors


def main() -> int:
    targets = _find_json_files()
    errors = _validate_files(targets)
    if errors:
        print(f"JSON validation failed for {len(errors)} file(s):")
        for path, message in errors:
            print(f" - {path}: {message}")
        return 1
    print(f"Validated {len(targets)} JSON file(s) successfully.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
